### PR TITLE
Scripts/Spells: Dancing Rune Weapon

### LINKED ISSUE #6192: [Core/Spells] Dancing Rune Weapon
Below may be not valid anymore, here's more fresh description of what is broken here https://github.com/TrinityCore/TrinityCore/issues/21538 + https://github.com/TrinityCore/TrinityCore/issues/25675
https://github.com/TrinityCore/TrinityCore/issues/6192#issuecomment-109767527 and in comments below
Abandoned PRs https://github.com/TrinityCore/TrinityCore/pull/16281 https://github.com/TrinityCore/TrinityCore/pull/23191

--

This spell was recently ( 24b699d904aa ) fixed, but I am getting these log errors:

> 2012-04-13 08:31:43 ERROR: Unit::HandleDummyAuraProc: Spell 49028 has non-existing triggered spell 1206
> 2012-04-13 08:31:43 ERROR: Unit::HandleDummyAuraProc: Spell 49028 has non-existing triggered spell 1206
> ...

Using hash 4daafd0
##

Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/64925-core-spells-dancing-rune-weapon?utm_campaign=plugin&utm_content=tracker%2F1310&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F1310&utm_medium=issues&utm_source=github).

### PR DESCRIPTION
**Pull request by Sorikoff**
Beginnings of attempt to implement Dancing Rune Weapon. This PR is based on ccrs work.

**Target branch(es):** 3.3.5

**Issues addressed:** Closes #6192

**Tests performed:** Works. Feedback is very welcome.

### CODE REVIEW

**Comment by illfated** [comment]
Happy to see that you have started working on it. Maybe DRW will become fully available in 2020? 😆

File: src/server/scripts/Pet/pet_dk.cpp [hunk 1]
```diff
@@ -137,6 +145,81 @@ class npc_pet_dk_guardian : public CreatureScript
...
+
+        me->GetThreatManager().RegisterRedirectThreat(SPELL_DK_DANCING_RUNE_WEAPON, summoner->GetGUID(), 100);
+    }
+
+    void SetGUID(ObjectGuid const& guid, int32 id) override
+    {
+        if (id == DATA_INITIAL_TARGET_GUID)
+        {
+            _targetGUID = guid;
+            if (Unit *target = ObjectAccessor::GetUnit(*me, _targetGUID))
```
Comment (illfated) [inline src/server/scripts/Pet/pet_dk.cpp]:
```cpp
src/server/scripts/Pet/pet_dk.cpp:181:32:
 fatal error: use of undeclared identifier 'ObjectAccessor'
            if (Unit *target = ObjectAccessor::GetUnit(*me, _targetGUID))
                               ^
```
This is a nopch error, so the solution is most likely to add
```cpp
#include "ObjectAccessor.h"
```
in the list of includes on top of the file.

**Comment by Sorikoff** [comment]
I don't think this spell is going to be workings retail-like. Looks to me that there is only few information regarding DRW ;(

**Comment by illfated** [comment]
That would be a reasonable explanation for why we have been without DRW for so long.

File: sql/updates/world/3.3.5/9999_99_99_99_world.sql [hunk 2]
```diff
@@ -0,0 +1,4 @@
+UPDATE `creature_template` SET `unit_flags`=`unit_flags`|33554434,`flags_extra`=0,`ScriptName`='npc_pet_dk_rune_weapon' WHERE `entry`=27893;
+DELETE FROM `spell_script_names` WHERE `ScriptName` = 'spell_dk_dancing_rune_weapon';
+INSERT INTO `spell_script_names` (`spell_id`, `ScriptName`) VALUES
+(49028, 'spell_dk_dancing_rune_weapon');
```
Comment (illfated) [inline sql/updates/world/3.3.5/9999_99_99_99_world.sql]: BTW, add a NewLine (CR/LF) at the end here.

**Comment by Sorikoff** [comment]
Right now DRW is kinda working. It follows owner in/out of combat and attacks its current target (ignoring crowd controlled). Visuals are also present. I applied scaling which was already present in TC, though data might not be precise enough, but damage looks fine to me.

File: src/server/scripts/Spells/spell_pet.cpp [hunk 3]
```diff
@@ -1697,67 +1697,53 @@ class spell_dk_pet_scaling_03 : public SpellScriptLoader
     }
 };
 
-class spell_dk_rune_weapon_scaling_02 : public SpellScriptLoader
+class spell_dk_rune_weapon_scaling_02 : public AuraScript
```
Comment (ccrs) [inline src/server/scripts/Spells/spell_pet.cpp]: DANGER

ElectricGregariousCivet-small
(lets see who knows this **amazing** song)

IIRC this spell is also used on dk's ghoul

File: src/server/scripts/Spells/spell_dk.cpp [hunk 4]
```diff
@@ -734,66 +735,66 @@ class spell_dk_corpse_explosion : public SpellScriptLoader
...
+        for (Creature* temp : runeWeapons)
+            temp->AI()->SetGUID(GetTarget()->GetGUID(), DATA_INITIAL_TARGET_GUID);
+    }
 
-                if (!drw || !drw->GetVictim())
-                    return;
+    bool CheckProc(ProcEventInfo& eventInfo)
+    {
+        if (SpellInfo const* procSpell = eventInfo.GetSpellInfo())
+            if (procSpell->IsRankOf(sSpellMgr->GetSpellInfo(SPELL_DK_BLOOD_STRIKE)) ||
```
Comment (ccrs) [inline src/server/scripts/Spells/spell_dk.cpp]: This is not the most efficient way of doing this. Pretty sure u can play with values on `spell_proc` table and get the same result.

**Comment by ccrs** [comment]
I think the best approach is this one:

1. I modify my PR to include everything BUT the scaling part
2. You struggle for a couple months to find the best solution for the scaling part
3. Maybe if I have some time to spare (OMEGALUL) I help you

File: src/server/scripts/Spells/spell_pet.cpp (hunk 3 above)
Comment (Sorikoff) [reply src/server/scripts/Spells/spell_pet.cpp]: Nice :rofl: Didn't notice that...

**Comment by Sorikoff** [comment]
About part 2. I think that even half-working (no proper scaling) DRW is better then not working at all. May be logic for this spell can implemented stand-alone? Stats will be corrected, when information will be found, if it ever will.

File: src/server/scripts/Spells/spell_pet.cpp (hunk 3 above)
Comment (Sorikoff) [reply src/server/scripts/Spells/spell_pet.cpp]: But that's AuraScript anyway.

Comment (Riztazz) [reply src/server/scripts/Spells/spell_pet.cpp]: @ccrs
Fire in the disco!

**Comment by Riztazz** [comment]
Spell 51906 reduces DRW damage done by 50% and since DRW is supposed to deal 50% of the DK damage, it should have exact copy of DK stats shouldnt it? And if i were to guess, scaling auras are the same as ghouls auras or gargoyle scaling auras (i dont think TC uses scaling auras for those tho)

File: src/server/scripts/Spells/spell_dk.cpp [hunk 5]
```diff
@@ -734,66 +735,66 @@ class spell_dk_corpse_explosion : public SpellScriptLoader
...
+            return;
 
-            void Register() override
-            {
-                OnEffectProc += AuraEffectProcFn(spell_dk_dancing_rune_weapon_AuraScript::HandleProc, EFFECT_1, SPELL_AURA_DUMMY);
-            }
-        };
+        for (Unit* controlled : GetTarget()->m_Controlled)
+            if (controlled->GetEntry() == NPC_DK_DANCING_RUNE_WEAPON)
+                controlled->CastSpell(controlled->GetVictim(), procSpell->Id);
```
Comment (Riztazz) [inline src/server/scripts/Spells/spell_dk.cpp]: this will fail if DRW is currently not attacking anything. EventInfo should be used for target selection?

**Comment by Riztazz** [comment]
You probably want to prevent that effect from firing as well:
```
Effect 2: Id 6 (SPELL_EFFECT_APPLY_AURA)
BasePoints = 10
Targets (6, 0) (TARGET_UNIT_TARGET_ENEMY, NO_TARGET)
Aura Id 4 (SPELL_AURA_DUMMY), value = 10, misc = 0 (0), miscB = 0, periodic = 0
```
When aura procs, this will procs as well and cause errors in console

File: src/server/scripts/Spells/spell_dk.cpp (hunk 4 above)
Comment (Riztazz) [reply src/server/scripts/Spells/spell_dk.cpp]: And if you stick to this approach, it lacks a few spells. Rune strike is one of them

**Comment by Sorikoff** [comment]
Not going to continue on this.

**Pull request closed without merging**